tensorflow/python/tools/saved_model_cli.py

```diff
@@ -137,24 +146,121 @@ def _show_inputs_outputs(saved_model_dir, tag_set, signature_def_key, indent=0):
+  """
+  if context.executing_eagerly():
+    # Disable eager execution to prevent loading of checkpoints
+    ops_lib.disable_eager_execution()
+  trackable_object = load.load(saved_model_dir)
+  indent_str = '  ' * indent
+
+  def in_print(s):
+    print(indent_str + s)
+
+  print('Defined Functions:')
+  functions = save._AugmentedGraphView(
+      trackable_object).list_functions(trackable_object)
+  for name, function in functions.items():
+    in_print('Function Name: \'%s\'' % name)
+    for index, concrete_functions in enumerate(
+            function._list_all_concrete_functions_for_serialization(), 1):
+      args, kwargs = (concrete_functions.structured_input_signature)
+      in_print('Option #%d' % index)
+      in_print('  Callable with:')
+      _print_args(args, indent=3)
+      if kwargs:
+        _print_args(kwargs, "Named Argument", indent=3)
+
+
+def _print_args(arguments, argument_type="Argument", indent=0):
+  """Formats and prints the argument of the concrete functions defined in the model
+
+     Args:
+       arguments: Arguments of the concrete functions.
+       argument_type: Type of Argument List to Format and print.
+       indent: How far (in increments of 2 spaces) to indent each line of output.
+  """
+  indent_str = '  ' * indent
+
+  def quotes(value):
+    is_quotes = '\'' * isinstance(value, str)
+    return is_quotes + value + is_quotes
+
+  def in_print(s, end='\n'):
+    print(indent_str + s, end=end)
+
+  def is_nested(args):
+    return nest.is_nested(args) and not isinstance(args, dict)
+  if is_nested(arguments):
+    for index, element in enumerate(arguments, 1):
+      if indent == 3:
+        in_print('%s #%d' % (argument_type, index))
+      if isinstance(element, tensor_spec.TensorSpec):
+        _print_tensor_spec(element, indent)
+      elif is_nested(element):
+        in_print('  DType: %s' % type(element).__name__)
+        in_print('  Values: [', end='')
+        _print_args(element, indent + 1)
+        in_print('  ]')
+      elif isinstance(element, dict):
+        in_print('  DType: %s' % type(element).__name__)
+        in_print('  Values:  {', end='')
+        for key, value in element.items():
+          if is_nested(element):
+            in_print('      \'%s\': [' % str(key), end='')
+            _print_args(element, indent + 1)
+            in_print('        ]')
+          else:
+            in_print('      \'%s\': %s' % (str(key), quotes(value)), end='')
+        in_print('      }')
+      else:
+        in_print('  DType: %s' % type(element).__name__)
+        in_print('  Value: %s' % str(element))
+
+
+def _print_tensor_spec(tensor_spec, indent=0):
+  """Prints details of the given tensor_spec.
+
+     Args:
+       tensor_spec: TensorSpec object to be printed.
+       indent: How far (in increments of 2 spaces) to indent each line output
+  """
+  indent_str = '  ' * indent
+
+  def in_print(s):
+    print(indent_str + s)
+  in_print(
```

Comment:
Why can't we use the native https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/framework/tensor_spec.py#L103? Basically print(tensor_spec) or repr(tensor_spec).

Reply:
We obviously can. But in reality most of the newbies would have no idea about TensorSpec, even I had no idea, before GSOC. So, if a user downloads a model from TFHub and tries to check with saved_model_cli, so I thought maybe it's better to show "Tensor", which is a basic requirement of understanding for TensorFlow, instead of "TensorSpec". What do you suggest?

Reply:
I am really not getting the educational point here. If we present things differently, it will just cause more confusion long term. Please remove this redundant code.

Reply:
> I am really not getting the educational point here. If we present things differently, it will just cause more confusion long term. Please remove this redundant code.

Okay! :)